torch/distributed/_shard/checkpoint/state_dict_saver.py

```diff
@@ -0,0 +1,159 @@
+import io
+from typing import Any, Dict, List, Tuple, Optional
+
+import torch
+import torch.distributed as dist
+
+from torch import Tensor
+from torch.distributed._shard.sharded_tensor import (
+    ShardedTensor,
+)
+
+from .metadata import (
+    Metadata,
+    BytesWriteRequest,
+    TensorWriteRequest,
+)
+from .resharding import (
+    _prepare_sharded_tensor_write,
+    _prepare_tensor_write,
+    _prepare_bytes_write
+)
+
+from .storage import StorageWriter
+
+# -------------- private functions --------------
+
+def _prepare(
+    state_dict: Dict[str, Any]
+) -> Tuple[Metadata, Dict[str, int], List[BytesWriteRequest], List[TensorWriteRequest]]:
+    """
+    Build the serialization plan for a given state_dict
+
+    Args:
+        state_dict: The instance to plan for.
+
+    Returns:
+        A tuple with the following values:
+
+        metadata: Metadata
+        The storage metadata describing Tensor and ShardedTensors
+        instances found in `state_dict`. See `Metadata` for the schema.
+
+        size_for_storage_keys: Dict[str, int]
+            Key is the storage key name, value is the associated size
+            It can used to pre allocate the storage for parallel and non sequential writes.
+
+        bytes_write_requests: List[BytesWriteRequest]
+            List of ByteIO write requests that should be performed by the writer.
+
+        tensor_write_requests: List[TensorWriteRequest]
+            List of Tensor write requests that should be performed by the writer.
+
+    """
+    metadata = Metadata(state_dict_metadata={})
+    tensor_write_requests: List[TensorWriteRequest] = []
+    bytes_write_requests: List[BytesWriteRequest] = []
+    storage_key_to_fqn: Dict[str, str] = dict()
+    # The assumption is that all non ShardedTensor items are replicated
+    #   and we can save them from rank 0.
+    write_replicated_data = not (dist.is_initialized() and dist.get_rank() != 0)
+
+    for fqn, obj in state_dict.items():
+        if isinstance(obj, ShardedTensor):
+            st_write_reqs, st_md = _prepare_sharded_tensor_write(obj, fqn, storage_key_to_fqn)
+            tensor_write_requests += st_write_reqs
+            metadata.state_dict_metadata[fqn] = st_md
```

Comment:
I'm assuming we plan to change this where only rank 0 computes the global metadata and not every rank?

Reply:
Yes, the plan is to gather each rank metadata after writes.